sys/kern/lockdep.c

```diff
@@ -0,0 +1,254 @@
+#if LOCKDEP
+
+#include <sys/lockdep.h>
+#include <sys/spinlock.h>
+#include <sys/thread.h>
+#include <sys/mimiker.h>
+#include <sys/klog.h>
+
+typedef TAILQ_HEAD(, lock_class) lock_class_list_t;
+
+typedef struct lock_class {
+  TAILQ_ENTRY(lock_class) hash_entry;
+  lock_class_key_t *key;
+  const char *name;
+  TAILQ_HEAD(, lock_class_link) locked_after;
+
+  int bfs_visited;
+} lock_class_t;
+
+/* lock_class_link represents an edge in the graph of lock depedency. */
+typedef struct lock_class_link {
+  TAILQ_ENTRY(lock_class_link) entry;
+  lock_class_t *from, *to;
+} lock_class_link_t;
+
+static spin_t main_lock = SPIN_INITIALIZER(0);
+
+#define CLASSHASH_SIZE 64
+#define CLASSHASH(key) ((uintptr_t)(key) % CLASSHASH_SIZE)
+#define CLASS_HASH_CHAIN(key) (&lock_hashtbl[CLASSHASH(key)])
+
+static lock_class_list_t lock_hashtbl[CLASSHASH_SIZE];
+
+#define MAX_CLASSES 64
+static lock_class_t lock_classes[MAX_CLASSES];
+static int class_cnt = 0;
+
+#define MAX_LINKS 256
+static lock_class_link_t lock_links[MAX_LINKS];
+static int link_cnt = 0;
+
+#define BFS_QUEUE_SIZE 32
+
+typedef struct {
+  lock_class_link_t *items[BFS_QUEUE_SIZE];
+  size_t head, tail;
+} bfs_queue_t;
+
+static bfs_queue_t bfs_queue;
+
+static inline void bfs_q_init(bfs_queue_t *q) {
+  q->head = q->tail = 0;
+}
+
+static inline int bfs_q_empty(bfs_queue_t *q) {
+  return q->head == q->tail;
+}
+
+static inline int bfs_q_full(bfs_queue_t *q) {
+  return ((q->tail + 1) % BFS_QUEUE_SIZE) == q->head;
+}
+
+static inline int bfs_q_enqueue(bfs_queue_t *q, lock_class_link_t *elem) {
+  if (bfs_q_full(q))
+    return 0;
+
+  q->items[q->tail] = elem;
+  q->tail = (q->tail + 1) % BFS_QUEUE_SIZE;
+  return 1;
+}
+
+static inline lock_class_link_t *bfs_q_dequeue(bfs_queue_t *q) {
+  lock_class_link_t *elem;
+
+  if (bfs_q_empty(q))
+    return NULL;
+
+  elem = q->items[q->head];
+  q->head = (q->head + 1) % BFS_QUEUE_SIZE;
+  return elem;
+}
+
+static inline void lockdep_lock(void) {
+  spin_lock(&main_lock);
+}
+
+static inline void lockdep_unlock(void) {
+  spin_unlock(&main_lock);
+}
+
+static void bfs_init(void) {
+  for (int i = 0; i < class_cnt; i++) {
+    lock_classes[i].bfs_visited = 0;
```

Comment:
Could we optimize this a bit? For example, let's use a generation number that is incremented with each call to `check_path()`, and a node is considered visited if its `bfs_visited` value is equal to the generation number. Then we won't need to iterate over all lock classes here.

Reply:
Actually, I thought of this, but since we won't traverse the graph very often I decided not to do that for the sake of simplicity.

But you have noted that, so I will change it.